Turner.Infrastructure.Mediator/Decorators/TransactionHandler.cs

```diff
@@ -0,0 +1,75 @@
+﻿using System;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Infrastructure;
+
+namespace Turner.Infrastructure.Mediator.Decorators
+{
+    public class TransactionBaseHandler<TRequest, TResult> where TResult : Response, new()
+    {
+        private readonly DbContext _context;
+
+        public TransactionBaseHandler(DbContext context)
+        {
+            _context = context;
+        }
+
+        public async Task<TResult> HandleAsync(TRequest request, Func<Task<TResult>> processRequest)
+        {
+            if (_context.Database.CurrentTransaction != null)
+            {
+                return await processRequest();
+            }
+            
+            using (var transaction = _context.Database.BeginTransaction())
+            {
+                var result = await processRequest();
+                if (result.HasErrors)
+                {
+                    transaction.Rollback();
+                    return result;
+                }
+
+                await _context.SaveChangesAsync();
+                transaction.Commit();
+                return result;
+            }
+        }
+    }
+
+    public class TransactionHandler<TRequest> : IRequestHandler<TRequest> where TRequest : IRequest
+    {
+        private readonly Func<IRequestHandler<TRequest>> _decorateeFactory;
+        private readonly TransactionBaseHandler<TRequest, Response> _transactionHandler;
+
+        public TransactionHandler(Func<IRequestHandler<TRequest>> decorateeFactory,
+            TransactionBaseHandler<TRequest, Response> transactionHandler)
+        {
+            _decorateeFactory = decorateeFactory;
```

Comment:
what is this 0_o

Reply:
http://simpleinjector.readthedocs.io/en/latest/aop.html#decoration

They use `decoratee`. What specifically doesn't make sense?